pyuvdata/uvdata/ms.py

```diff
@@ -126,11 +754,303 @@ def find_not_ascii(s):
+
+        if run_check:
+            self.check(
+                check_extra=check_extra,
+                run_check_acceptability=run_check_acceptability,
+                strict_uvw_antpos_check=strict_uvw_antpos_check,
+            )
+
+        if os.path.exists(filepath):
+            if clobber:
+                print("File exists; clobbering")
+            else:
+                raise IOError("File exists; skipping")
+
+        # TODO: the phasing requirement can probably go away once we have proper
+        # multi-object support.
+        if self.phase_type == "phased":
+            pass
+        elif self.phase_type == "drift":
+            if force_phase:
+                print(
+                    "The data are in drift mode and do not have a "
+                    "defined phase center. Phasing to zenith of the first "
+                    "timestamp."
+                )
+                phase_time = Time(self.time_array[0], format="jd")
+                self.phase_to_time(phase_time)
+            else:
+                raise ValueError(
+                    "The data are in drift mode. "
+                    "Set force_phase to true to phase the data "
+                    "to zenith of the first timestamp before "
+                    "writing a uvfits file."
+                )
+        else:
+            raise ValueError(
+                "The phasing type of the data is unknown. "
+                "Set the phase_type to drift or phased to "
+                "reflect the phasing status of the data"
+            )
+
+        # Initialize a skelton measurement set
+        ms = self._init_ms_file(filepath)
+
+        # All all the rows we need up front, which will allow us to fill the columns
+        # all in one shot.
+        ms.addrows(self.Nblts * self.Nspws)
+
+        attr_list = ["data_array", "nsample_array", "flag_array"]
+        col_list = ["DATA", "WEIGHT_SPECTRUM", "FLAG"]
+
+        # Some tasks in CASA require a band-representative (band-averaged?) value for
+        # the weights and noise for all channels in each row in the MAIN table, which
+        # we will roughly calculate in temp_weights below.
+        temp_weights = np.zeros((self.Nblts * self.Nspws, self.Npols), dtype=float)
+
+        for attr, col in zip(attr_list, col_list):
+            val_dict = {}
+            temp_array = getattr(self, attr)
+
+            if not self.future_array_shapes:
+                temp_array = np.squeeze(temp_array, axis=1)
 
-    # ms write functionality to be added later.
-    def write_ms(self):
-        """Write ms: Not yet supported."""
+            for idx, spw_id in enumerate(self.spw_array):
+                if self.flex_spw:
+                    sel_mask = self.flex_spw_id_array == spw_id
+                else:
+                    sel_mask = np.ones(self.Nfreqs, dtype=bool)
+
+                temp_dict = {
+                    "r%i" % ((self.Nspws * jdx) + 1): temp_array[None, jdx, sel_mask, :]
```

Comment:
@kartographer -- Can you double check if this is right? That should produce 8 SPWs in the output MS. Otherwise, the dictionary key is the same for all SPWs.
```suggestion
                    "r%i" % ((self.Nspws * idx) + 1): temp_array[None, jdx, sel_mask, :]
```

Reply:
Whoops -- yeah, the above was missing a piece of the indexing scheme (should have been `(self.Nspws * jdx) + 1 + idx`), can you see if it works now?